java/ql/lib/semmle/code/java/dataflow/ListOfConstantsSanitizer.qll

```diff
@@ -0,0 +1,357 @@
+/**
+ * Provides classes for identifying comparisons against a list of compile-time
+ * constants and considering them as taint-sanitizers.
+ */
+
+import java
+private import semmle.code.java.controlflow.Guards
+private import semmle.code.java.dataflow.TaintTracking
+private import internal.DataFlowPrivate
+
+/**
+ * A comparison against a list of compile-time constants, sanitizing taint by
+ * restricting to a set of known values.
+ */
+class ListOfConstantsComparisonSanitizerGuard extends TaintTracking::DefaultTaintSanitizer {
+  ListOfConstantsComparisonSanitizerGuard() {
+    this = DataFlow::BarrierGuard<listOfConstantsComparisonSanitizerGuard/3>::getABarrierNode()
+  }
+}
+
+private predicate listOfConstantsComparisonSanitizerGuard(Guard g, Expr e, boolean outcome) {
+  exists(ListOfConstantsComparison locc |
+    g = locc and
+    e = locc.getExpr() and
+    outcome = locc.getOutcome()
+  )
+}
+
+private Expr skipCaseChanges(MethodCall mc) {
+  exists(Method changecase | mc.getMethod() = changecase |
+    changecase.hasName(["toUpperCase", "toLowerCase"]) and
+    changecase.getDeclaringType() instanceof TypeString
+  ) and
+  result = mc.getQualifier()
+}
+
+/** A comparison against a list of compile-time constants. */
+abstract class ListOfConstantsComparison extends Guard {
+  Expr e;
+  boolean outcome;
+
+  ListOfConstantsComparison() {
+    exists(this) and
+    outcome = [true, false]
+  }
+
+  /**
+   * Gets the expression that is compared to a list of constants. Note that it
+   * is very common to see `x.toUpperCase()` or `x.toLowerCase()` compared with
+   * a list of constants, and in this case `result` is `x`.
+   */
+  Expr getExpr() {
+    result = skipCaseChanges(e)
+    or
+    not exists(skipCaseChanges(e)) and result = e
+  }
+
+  /** Gets the value of `this` when `e` is in the list of constants. */
+  boolean getOutcome() { result = outcome }
+}
+
+/**
+ * Holds if the method call `mc` has only compile-time constant arguments (and
+ * at least one argument). To account for varargs methods, we also include
+ * a single array argument which is initialized locally with at least one
+ * compile-time constant.
+ */
+predicate methodCallHasConstantArguments(MethodCall mc) {
+  // f("a", "b", "c")
+  forex(Expr e | e = mc.getAnArgument() |
+    not e.getType() instanceof Array and e.isCompileTimeConstant()
+  )
+  or
+  // String[] a = {"a", "b", "c"};
+  // f(a)
+  mc.getNumArgument() = 1 and
+  mc.getArgument(0).getType() instanceof Array and
+  exists(ArrayInit arr | DataFlow::localExprFlow(arr, mc.getArgument(0)) |
+    forex(Expr e | e = arr.getAnInit() | e.isCompileTimeConstant())
+  )
+}
+
+/** The name of a class implementing `java.util.Collection`. */
+class CollectionClass extends string {
+  CollectionClass() { this = ["List", "Set"] }
+}
+
+/** Classes for `java.util.List` and `java.util.Set`. */
+module Collection {
```

Comment:
This is not really a general module for collection classes, and if it were, it probably shouldn't be in a library called `ListOfConstantsSanitizer.qll`, hence make private.

Reply:
My thinking in not making this all private was that someone could add some classes representing their favourite way of making a list of constants.